docs: state-transition-function (#2133)
OffchainLabs/arbitrum-docs: arbitrum-docs/how-arbitrum-works/04-state-transition-function/03-ethereum-vs-arbitrum.mdx

---
title: 'State Transition Function: Ethereum vs. Arbitrum'
description: 'Learn the fundamentals of Nitro, Arbitrum stack.'
author: petevielhaber
sme: Mehdi
user_story: As a current or prospective Arbitrum user, I need learn more about Nitros design.
content_type: get-started
---

import ImageZoom from '@site/src/components/ImageZoom';

This section explains the architecture and implementation of the <a data-quicklook-from="state-transition-function">State Transition Function</a> (STF) in Ethereum and the <a data-quicklook-from="arbitrum">Arbitrum</a> Nitro stack, highlighting their similarities and key differences.

## STF in Ethereum

In Ethereum, the STF receives transactions as inputs, processes them via the EVM, and produces the final state as output.

> In Ethereum, the state is a vast data structure represented by a modified Merkle Patricia Trie. This structure holds all accounts, linking them via hashes and reducing the entire state to a single root hash stored on the <a data-quicklook-from="blockchain">blockchain</a>.

Reviewer: ```suggestion
The Ethereum state is a vast data structure represented by a modified Merkle Patricia Trie. This structure holds all accounts, linking them via hashes and reducing the entire state to a single root hash stored on the <a data-quicklook-from="blockchain">blockchain</a>.
```

> The Ethereum Virtual Machine (EVM) operates similarly to a mathematical function: given as an input, it produces a deterministic output. Ethereum's STF encapsulates this behavior:

Reviewer: ```suggestion
The Ethereum Virtual Machine (EVM) operates similarly to a mathematical function: given an input, it produces a deterministic output. Ethereum's STF encapsulates this behavior:
```

> `Y(S, T) = S'`

Reviewer 2: ```suggestion
$$
Y(S, T) = S'
$$
```

Here, `S` represents the current state, `T` denotes the <a data-quicklook-from="transaction">transaction</a>, and `S'` is the new state resulting from the execution of `T`.

> The design of EVM is a stack machine with a maximum depth of 1024 items. Each item is a 256-bit word, chosen for compatibility with 256-bit cryptography (e.g., Keccak-256 hashes and secp256k1 signatures).

Reviewer: ```suggestion
The EVM operates as a stack machine with a maximum depth of 1024 items. Each item is a 256-bit word, chosen for compatibility with 256-bit cryptography (e.g., Keccak-256 hashes and secp256k1 signatures).
```

During execution, the EVM uses transient _memory_ (a word-addresses byte array) that only persists for the duration of a transaction. In contrast, each contract maintains a persistent Merkle Patricia _storage_ trie–a word-addressable word array–that forms part of the global state.

<a data-quicklook-from="smart-contract">smart contract</a> bytecode compiles into a series of EVM opcodes
that perform standard stack operations (such as `XOR`, `AND`, `ADD`, `SUB`) and blockchain-specific operations

> (such as `ADDRESS`, `BALANCE`, `BLOCKHASH`).

Reviewer: Nice!

engine. It processes transactions by executing the smart contract's bytecode and updating the global
state, ensuring that every state change is deterministic and secure.

In essence, Geth converts transaction inputs into precise computational steps within the EVM, maintaining the intricate data structures that underpin Ethereum's blockchain. Its robust design not only powers the core operations of Ethereum but also provides the foundation for advanced modificaitons in platforms like the <a data-quicklook-from="arbitrum-nitro">Arbitrum Nitro</a> stack.

## STF on Arbitrum

The Arbitrum Nitro stack implements a modified version of Ethereum's STF. While it retains the core principles of Ethereum, several Arbitrum-specific features and processes distinguish it from Ethereum's implementation. Key differences include:

- **Gas accounting**: In Arbitrum, executing a transaction incurs two costs, one for executing on the <a data-quicklook-from="child-chain">child chain</a> and another for submitting transaction batches to the <a data-quicklook-from="parent-chain">parent chain</a>. This dual fee structure requires a different approach to gas accounting.

> - **Base fee mechanism**: Gas prices and base fees on Arbitrum change differently from Ethereum's. Arbitrum maintains two base fees: one to track the cost of execution on the child chain and another to estimate the cost of posting batches to the parent chain. The mechanism that changes the child chain basefee differs from what is currently on Ethereum.

Reviewer: ```suggestion
- **Base fee mechanism**: Gas prices and base fees on Arbitrum change differently from Ethereum's. Arbitrum maintains two base fees: one to track the cost of execution on the child chain and another to estimate the cost of posting batches to the parent chain. The mechanism that changes the child chain's base fee differs from what is currently on Ethereum.
```

- **Precompiles**: Specific precompiled contracts on the Arbitrum chain will process Arbitrum-specific features. Moreover, individual <a data-quicklook-from="arbitrum-orbit">Arbitrum Orbit</a> chains may implement their own custom precompiles.

- **Custom transaction types**: Nitro Geth introduces several child chain-specific transaction types that utilize the non-standard [EIP-2718: Typed Transaction Envelope](https://eips.ethereum.org/EIPS/eip-2718). These types represent different transactions on the Arbitrum chain, such as retryable transactions.

- **<a data-quicklook-from="stylus">Stylus</a> in the STF**: With the introduction of Stylus, Arbitrum Nitro now supports executing smart contracts written in Rust, C, and C++, offering a high-performance alternative to the EVM. Unlike standard Ethereum transactions, Stylus contracts execute in a <a data-quicklook-from="wasm">WASM</a>-based runtime within <a data-quicklook-from="arbos">ArbOS</a>, leveraging efficient host I/O operations and optimized caching mechanisms.

> 

Reviewer: Pretty great explanation.

Reviewer: The items of this list could be changed into markdown headers.

Author: Good call

## Implementation overview

The Arbitrum STF is implemented using a modified version of Geth and integrated with ArbOS. This combination allows Arbitrum to support additional functionalities while maintaining compatibility with Ethereum's core execution model. The following diagram provides an abstract overview:

> <ImageZoom src="/img/haw-geth-sandwich.svg" alt="Geth sandwich" className="img-600px" />

Reviewer: Yaaaay! Dia-grams! Dia-grams!

The State Transition Function (STF) itself consists of the modified Geth core (base layer) combined with a portion of ArbOS (middle layer). In practice, the STF is a designated function within the source code (including all the code it calls), which:

- Takes the bytes of a transaction received in the inbox as input.

- Operates on a modifiable copy of the Ethereum state tree.

- May modify the state during execution.

- Emits a new block header (formatted in Ethereum's block header structure) that appends to the Nitro chain.

## Modified Geth on Nitro

Nitro's design builds upon the robust foundation of Geth with targeted modifications that enable it to support child chain-specific functionalities. These modifications include:

- **Custom hooks and interface adaptions**: Nitro integrates a series of custom hooks within Geth's transaction processing flow. These hooks allow Nitro to implement dual gas accounting, manage parent chain calldata fees, and handle custom transaction types–such as retryable transactions–without deviating from Ethereum's core logic.

> 

Reviewer: Would a link to the source code make sense here?

Author: Will add link to the Nitro repo
